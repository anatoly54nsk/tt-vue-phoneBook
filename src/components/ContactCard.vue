<template>
    <md-card class="card">
        <md-card-header>
            <md-card-header-text>
                <div class="md-title">{{name}}</div>
            </md-card-header-text>
        </md-card-header>
        <md-card-content>
            <div>
                <md-icon class="md-accent">phone</md-icon>
                <span class="field">{{contact.phone}}</span>
            </div>
            <div>
                <md-icon class="md-accent">email</md-icon>
                <span class="field">{{contact.email}}</span>
            </div>
        </md-card-content>
        <md-card-actions>
            <md-button @click="edit" class="md-accent contact-card__button_edit">Edit</md-button>
        </md-card-actions>
        <md-card-actions>
            <md-button @click="remove" class="md-accent contact-card__button_remove">Delete</md-button>
        </md-card-actions>
    </md-card>
</template>

<script>
    export default {
        name: "ContactCard",
        computed: {
            name() {
                return this.contact.name ? this.contact.name : this.contact.phone;
            },
        },
        props: {
            contact: {
                required: true,
            },
        },
        methods: {
            edit() {
                this.$emit('edit', this.contact.id);
            },
            remove() {
                this.$emit('remove', this.contact.id);
            },
        },
    }
</script>

<style scoped>
    .card {
        width: 250px;
    }

    .field {
        margin-left: 15px;
    }
</style>
