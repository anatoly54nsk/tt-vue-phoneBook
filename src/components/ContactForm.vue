<template>
    <div>
        <form novalidate class="md-layout md-alignment-center form" @submit.prevent>
            <md-card class="md-layout-item md-size-50 md-small-size-100">
                <md-card-header>
                    <div class="md-title">Contact</div>
                </md-card-header>

                <md-card-content>
                    <div class="md-layout md-gutter">
                        <div class="md-layout-item md-size-100">
                            <md-field :class="getValidationClass('name')">
                                <label for="name">Имя</label>
                                <md-input name="name" id="name" autocomplete="given-name"
                                          v-model="contact.name" :disabled="sending"/>
                                <span class="md-error"
                                      v-if="!$v.contact.name.required">Поле Имя обязательно для ввода</span>
                                <span class="md-error"
                                      v-else-if="!$v.contact.name.minLength">Длинна поля Имя должна быть не меньше 3</span>
                            </md-field>
                        </div>
                    </div>

                    <div class="md-layout md-gutter">
                        <div class="md-layout-item md-size-100">
                            <md-field :class="getValidationClass('phone')">
                                <label for="phone">Телефон</label>
                                <md-input type="text" id="phone" name="phone" autocomplete="phone"
                                          v-model="contact.phone"
                                          @keyup="checkPhone()"
                                          :disabled="sending"/>
                                <span class="md-error"
                                      v-if="!$v.contact.phone.required">The phone is required</span>
                                <span class="md-error" v-else-if="!$v.contact.phone.pattern">Invalid phone 3</span>
                            </md-field>
                        </div>
                    </div>

                    <div class="md-layout md-gutter">
                        <div class="md-layout-item md-size-100">
                            <md-field :class="getValidationClass('email')">
                                <label for="email">Email</label>
                                <md-input type="email" name="email" id="email" autocomplete="email"
                                          v-model="contact.email"
                                          :disabled="sending"/>
                                <span class="md-error"
                                      v-if="!$v.contact.email.required">The email is required</span>
                                <span class="md-error" v-else-if="!$v.contact.email.email">Invalid email</span>
                            </md-field>
                        </div>
                    </div>
                </md-card-content>

                <md-progress-bar md-mode="indeterminate" v-if="sending"/>

                <md-card-actions v-if="formMode === 'update'">
                    <md-button type="submit" class="md-accent" :disabled="sending" @click="update">Обновить</md-button>
                </md-card-actions>

                <md-card-actions v-if="formMode === 'create'">
                    <md-button type="submit" class="md-accent" :disabled="sending" @click="add">Добавить</md-button>
                </md-card-actions>
            </md-card>
        </form>
    </div>
</template>


<script>
    import {validationMixin} from 'vuelidate'
    import {
        required,
        email,
        minLength,
    } from 'vuelidate/lib/validators'

    const phonePattern = /^(\+{1}\d{1,3})\s?-?\(?(\d{3})\)?\s?-?(\d{3})\s?-?(\d{4})$/;

    export default {
        name: "ContactForm",
        mixins: [validationMixin],
        props: {
            contact: null,
            formMode: null,
            contactSaved: null,
            sending: null,
        },
        validations: {
            contact: {
                name: {
                    required,
                    minLength: minLength(3),
                },
                phone: {
                    required,
                    pattern: value => {
                        const phoneParts = value.match(phonePattern);
                        return phoneParts !== null;
                    },
                },
                email: {
                    required,
                    email,
                },
            },
        },
        methods: {
            getValidationClass(fieldName) {
                const field = this.$v.contact[fieldName];

                if (field) {
                    return {
                        'md-invalid': field.$invalid && field.$dirty
                    }
                }
            },
            add() {
                if (this.validateForm()) {
                    const contact = {...this.contact};
                    contact.phone = this.preparePhone(contact.phone);
                    this.$v.$reset();
                    this.$emit('add', contact);
                }
            },
            update() {
                if (this.validateForm()) {
                    const contact = {...this.contact};
                    contact.phone = this.preparePhone(contact.phone);
                    this.$v.$reset();
                    this.$emit('update', contact);
                }
            },
            preparePhone(phone) {
                let phoneParts = phone.match(phonePattern);
                phoneParts.shift();
                return `${phoneParts[0]} ${phoneParts[1]} ${phoneParts[2]} ${phoneParts[3]}`;
            },
            validateForm() {
                this.$v.$touch();
                return !this.$v.$invalid;
            },
            checkPhone() {
                if (this.contact.phone) {
                    let lastLetter = this.contact.phone.slice(-1);
                    let matches = lastLetter.match(/^\d?\(?\)?-?\+?\s?$/g);
                    if (!matches) {
                        this.undoPhoneEntering();
                    }
                }
            },
            undoPhoneEntering() {
                this.contact.phone = this.contact.phone.slice(0, -1);
            },
        },
        beforeRouteUpdate() {
            this.destroy();
        },
    }
</script>

<style scoped>
    .form {
        padding: 50px 15px;
    }

    .md-focused label {
        color: #ff5252 !important;
    }

    .md-focused input {
        color: #ff5252 !important;
    }

    .md-focused::after, .md-focused::before, .md-field::before {
        background-color: #ff5252;
    }
</style>
