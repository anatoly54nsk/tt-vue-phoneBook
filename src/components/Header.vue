<template>
    <md-toolbar class="md-accent" md-elevation="1">
        <h2 class="md-title" style="flex: 1">{{title}}</h2>
        <md-button class="md-icon-button md-raised" @click="home" :disabled="currentPage === 'home'">
            <md-icon>home</md-icon>
            <md-tooltip md-direction="top">На главную</md-tooltip>
        </md-button>
        <md-button class="md-icon-button md-raised" @click="add" :disabled="currentPage === 'add'">
            <md-icon>add</md-icon>
            <md-tooltip md-direction="top">Добавить контакт</md-tooltip>
        </md-button>
        <md-button class="md-icon-button md-raised" @click="refresh">
            <md-icon>refresh</md-icon>
            <md-tooltip md-direction="top">Обновить</md-tooltip>
        </md-button>
    </md-toolbar>
</template>

<script>
    export default {
        name: "Header",
        props: {
            title: {
                required: true,
            },
            currentPage: {
                required: true,
            },
        },
        methods: {
            home() {
                this.$emit('home');
            },
            add() {
                this.$emit('add');
            },
            refresh() {
                this.$emit('refresh');
            },
        },
    }
</script>

<style scoped>

</style>
