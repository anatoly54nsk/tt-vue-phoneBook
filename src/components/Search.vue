<template>
    <md-card-content>
        <div class="md-layout md-gutter">
            <div class="md-layout-item md-size-100">
                <md-field>
                    <label for="search">Поиск</label>
                    <md-input id="search" v-model="query" @keyup="search"/>
                </md-field>
            </div>
        </div>
    </md-card-content>
</template>

<script>
    export default {
        name: "Search",
        data() {
            return {
                query: '',
            }
        },
        methods: {
            search() {
                this.$emit('search', this.query);
            },
        },
        created() {
            this.$eventHub.$on('refresh', () => this.query = '');
        }
    }
</script>

<style scoped>
    .md-focused label {
        color: #ff5252 !important;
    }

    .md-focused input {
        color: #ff5252 !important;
    }

    .md-focused::after, .md-focused::before, .md-field::before {
        background-color: #ff5252 !important;
    }
</style>
