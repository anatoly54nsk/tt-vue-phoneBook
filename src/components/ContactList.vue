<template>
    <div class="content">
        <h1 class="no-variants" v-if="list && list.length===0">Нет похожих контактов</h1>
        <div class="contact-list">
            <ContactCard v-for="contact in list"
                         :contact="contact"
                         :key="contact.id"
                         @edit="edit"
                         @remove="remove">
            </ContactCard>
        </div>
    </div>
</template>

<script>
    import ContactCard from "./ContactCard";

    export default {
        name: "ContactList",
        components: {
            ContactCard
        },
        props: {
            list: Array,
        },
        methods: {
            edit(id) {
                this.$emit('edit', id);
            },
            remove(id) {
                this.$emit('remove', id);
            },
        },
    }
</script>

<style scoped>
    .content {
        width: 100%;
    }

    .contact-list {
        padding: 15px 0;
        width: 100%;
        display: grid;
        grid-template-columns: 25% 25% 25% 25%;
        justify-items: center;
        grid-row-gap: 15px;
    }

    .no-variants {
        display: block;
        text-align: center;
        width: 100%;
        color: #ff5252;
    }
</style>
