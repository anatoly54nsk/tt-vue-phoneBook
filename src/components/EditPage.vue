<template>
    <div class="page">
        <ContactForm
                :contact-saved="contactSaved"
                :formMode="formMode"
                :sending="sending"
                :contact="contact"
                @add="add"
                @update="update"></ContactForm>
    </div>
</template>

<script>
    import contactApi from "../services/HerokuApi";
    import ContactForm from "./ContactForm";

    export default {
        name: "EditPage",
        components: {
            ContactForm
        },
        data() {
            return {
                contactSaved: true,
                sending: false,
                formMode: 'create',
                contact: {
                    id: null,
                    name: null,
                    phone: null,
                    email: null,
                },
            }
        },
        methods: {
            add(contact) {
                contactApi.create(contact)
                    .then(() => this.afterSave());
            },
            update(contact) {
                contactApi.update(contact)
                    .then(() => this.afterSave());
            },
            afterSave() {
                this.contactSaved = true;
                this.sending = false;
                this.$eventHub.$emit('refresh');
                this.$router.push('/');
            },
        },
        created() {
            if (this.$router.currentRoute.params.id) {
                contactApi.get(this.$router.currentRoute.params.id)
                    .then(contact => this.contact = contact);
                this.formMode = 'update';
            }
        },
    }
</script>

<style scoped>

</style>
