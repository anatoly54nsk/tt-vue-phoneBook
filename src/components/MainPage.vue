<template>
    <div class="page">
        <Search @search="filterList"></Search>
        <ContactList v-if="!updating" :list="filteredContactList" @edit="edit" @remove="remove"></ContactList>
        <div class="spinner">
            <md-progress-spinner v-if="updating" md-mode="indeterminate"  class="md-accent"></md-progress-spinner>
        </div>
    </div>
</template>

<script>
    import ContactsApi from "../services/ContactsApi";
    import ContactList from "./ContactList";
    import Search from "./Search";

    export default {
        name: "MainPage",
        components: {
            ContactList,
            Search,
        },
        data() {
            return {
                contactList: {},
                filteredContactList: {},
                updating: false,
            }
        },
        methods: {
            update() {
                this.updating = true;
                ContactsApi.getContacts()
                    .then(list => {
                        this.updating = false;
                        this.contactList = list;
                        this.filterList();
                    });
            },
            filterList(query = null) {
                this.filteredContactList = query ?
                    this.contactList.filter(contact => Object.values(contact).join('').toLowerCase().includes(query.toLowerCase()))
                    : this.contactList;
            },
            edit(id) {
                window.console.log(id);
                this.$router.push(`/edit/${id}`);
            },
            remove(id) {
                ContactsApi.deleteContact(id).then(
                    () => this.update()
                );
            },
        },
        created() {
            this.$eventHub.$on('refresh', this.update);
            this.update();
        },
        beforeDestroy() {
            this.$eventHub.$off('refresh');
        },
    }
</script>

<style scoped>
    .spinner {
        display: flex;
        justify-content: center;
        width: 100%;
    }
</style>
