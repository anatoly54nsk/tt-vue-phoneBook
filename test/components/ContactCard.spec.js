import ContactCard from "../../src/components/ContactCard";
import {createLocalVue, mount} from "@vue/test-utils";
import {MdButton, MdCard, MdIcon} from "vue-material/dist/components";
import Vuetify from 'vuetify';
import Vue from 'vue';

Vue.use(Vuetify);

describe('ContactCard', () => {

    let wrapper;
    let localVue;
    let vuetify;
    let validContact;
    let invalidContact;

    beforeEach(() => {
        vuetify = new Vuetify();
        localVue = createLocalVue();
        localVue.use(MdButton);
        localVue.use(MdCard);
        localVue.use(MdIcon);
        validContact = {
            id: 1,
            name: 'exampleName',
            phone: '+78923334444',
            email: 'example@example.exmpl'
        };
        invalidContact = {
            id: null,
            name: null,
            phone: '+73334444',
            email: 'exxample.exmpl'
        };
    });

    it('should render correctly', () => {
        wrapper = mount(ContactCard, {vuetify, localVue, propsData: {contact: validContact}});
        expect(wrapper.vm.$el).toMatchSnapshot();
    });

    it('should show name when contact.name exists', () => {
        wrapper = mount(ContactCard, {localVue, propsData: {contact: validContact}});
        expect(wrapper.vm.name).toBe(validContact.name);
    });

    it('should show phone when contact.name don`t exists', () => {
        wrapper = mount(ContactCard, {localVue, propsData: {contact: invalidContact}});
        expect(wrapper.vm.name).toBe(invalidContact.phone);
    });

    it('should called computed name function', () => {
        const fn = jest.fn();
        wrapper = mount(ContactCard, {localVue, computed: {name: fn}, propsData: {contact: validContact}});

        expect(fn).toBeCalledTimes(1);
    });

    it('should called only edit method on edit button click', () => {
        const editBtn = jest.fn();
        const removeBtn = jest.fn();
        wrapper = mount(ContactCard, {
            localVue,
            methods: {edit: editBtn, remove: removeBtn},
            propsData: {contact: validContact}
        });
        wrapper.find('button.contact-card__button_edit').trigger('click');

        expect(editBtn).toBeCalledTimes(1);
        expect(removeBtn).not.toBeCalled();
    });

    it('should called only remove method on remove button click', () => {
        const editBtn = jest.fn();
        const removeBtn = jest.fn();
        wrapper = mount(ContactCard, {
            localVue,
            methods: {edit: editBtn, remove: removeBtn},
            propsData: {contact: validContact}
        });
        wrapper.find('button.contact-card__button_remove').trigger('click');

        expect(removeBtn).toBeCalledTimes(1);
        expect(editBtn).not.toBeCalled();
    });

    it('should emit only edit event on edit button click', () => {
        wrapper = mount(ContactCard, {
            localVue,
            propsData: {contact: validContact}
        });
        wrapper.find('button.contact-card__button_edit').trigger('click');

        expect(wrapper.emitted('edit')).toBeTruthy();
        expect(wrapper.emitted('remove')).toBeFalsy();
    });

    it('should emit only remove event on remove button click', () => {
        wrapper = mount(ContactCard, {
            localVue,
            propsData: {contact: validContact}
        });
        wrapper.find('button.contact-card__button_remove').trigger('click');

        expect(wrapper.emitted('remove')).toBeTruthy();
        expect(wrapper.emitted('edit')).toBeFalsy();
    });
});
