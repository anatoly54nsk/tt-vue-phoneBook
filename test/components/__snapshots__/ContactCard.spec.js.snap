// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactCard should render correctly 1`] = `
<div
  class="md-card card md-theme-default"
>
  <div
    class="md-card-header md-card-header-flex"
  >
    <div
      class="md-card-header-text"
    >
      <div
        class="md-title"
      >
        exampleName
      </div>
    </div>
  </div>
   
  <div
    class="md-card-content"
  >
    <div>
      <i
        class="md-icon md-icon-font md-accent md-theme-default"
      >
        phone
      </i>
       
      <span
        class="field"
      >
        +78923334444
      </span>
    </div>
     
    <div>
      <i
        class="md-icon md-icon-font md-accent md-theme-default"
      >
        email
      </i>
       
      <span
        class="field"
      >
        example@example.exmpl
      </span>
    </div>
  </div>
   
  <div
    class="md-card-actions md-alignment-right"
  >
    <button
      class="md-button md-accent contact-card__button_edit md-theme-default"
      type="button"
    >
      <div
        class="md-ripple"
      >
        <div
          class="md-button-content"
        >
          Edit
        </div>
         
      </div>
    </button>
  </div>
   
  <div
    class="md-card-actions md-alignment-right"
  >
    <button
      class="md-button md-accent contact-card__button_remove md-theme-default"
      type="button"
    >
      <div
        class="md-ripple"
      >
        <div
          class="md-button-content"
        >
          Delete
        </div>
         
      </div>
    </button>
  </div>
</div>
`;
